### PR TITLE
[JENKINS-67744] make agent availability help more flexible

### PR DESCRIPTION
**Pull request by mawinter69**
each retention strategy now has it's own help file that is included automatically.

The old help files need to be kept for now as other plugins are still relying on them

See [JENKINS-67744](https://issues.jenkins-ci.org/browse/JENKINS-67744).

Currently nearly all implementations in plugins are for cloud purposes and do not even provide a descriptor so they are not presented in the UI at all.
What this will not fix is when plugins implement own extensions of node/slave and use in their jelly files the existing code with the hard coded link to the help for the retention strategy. One could create an own tag in lib/form for this eventually

### Proposed changelog entries

* Improve agent availability help.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by NotMyFault — COMMENTED** [review]
The double line breaks seem a bit unneeded, because they clutter the in-paragraph sentences quite a bit.

**Comment by mawinter69** [comment]
> The double line breaks seem a bit unneeded, because they clutter the in-paragraph sentences quite a bit.

I didn't change the formatting of the texts compared to what was there in the html. But that doesn't mean it has to stay like it

**Comment by NotMyFault** [comment]
> 

Yeah you already got it sorted, I found the couple of newlines weird to read through, especially the one breaking one paragraph.

Though it looks like GitHub broke the file encoding while I suggested the typo fix, sorry for that.

**Review by NotMyFault — APPROVED** [review]

File: core/src/main/resources/hudson/model/Slave/help-retentionStrategy.jelly
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/model/Slave/help-retentionStrategy.jelly:4]: Needs fix

File: core/src/main/resources/hudson/slaves/Messages_nl.properties
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/slaves/Messages_nl.properties:3]: Needs fix

File: core/src/main/resources/hudson/slaves/Messages_nl.properties
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/slaves/Messages_nl.properties:23]: slaafnode is "Slave Node"

**Review by oleg-nenashev — DISMISSED** [review]
"slaafnode" is "slave node". This is old terminology, and it must not be used in new logic

File: core/src/main/resources/hudson/slaves/Messages_nl.properties
Comment (mawinter69) [reply core/src/main/resources/hudson/slaves/Messages_nl.properties]: Replaced slaafnode with agent at all places.
Hope that this is a correct to use agent here in dutch

**Comment by timja** [comment]
Would be good to provide testing instructions for PRs like this.

I've created a permanent agent here:
http://localhost:8080/computer/createItem

Clicked the help icon next to 'Availability'

I get:
image

Would it be better to do this how pipeline syntax does it:

image

That way we don't have a huge help blob once plugins implement this?

Each one you choose can have more information added to it.

**Comment by mawinter69** [comment]
That is probably even better. Need to check how this is done in pipeline.
I looked up some of the other retention strategies and all were related to cloud agents and had no descriptor, so they wouldn't appear at all.

**Comment by timja** [comment]
> I looked up some of the other retention strategies and all were related to cloud agents and had no descriptor, so they wouldn't appear at all.

Seems the Azure ones do 😂 .

**Comment by mawinter69** [comment]
https://github.com/jenkinsci/jenkins/pull/5417
actually contains parts of the solution to use `f:dropdownDescriptorSelector` 
Then only the help files would be required

**Comment by mawinter69** [comment]
similar change could be done for the launch method

**Comment by mawinter69** [comment]
The change to the f:dropdownDescriptorSelector will have a negative impact to the ssh build agent plugin. There the help is manually added to each config.jelly file for the host key verification strategy.
There might be other places with the same problem

**Comment by timja** [comment]
> The change to the f:dropdownDescriptorSelector will have a negative impact to the ssh build agent plugin. There the help is manually added to each config.jelly file for the host key verification strategy. There might be other places with the same problem

Link?

PRs can be filed there to fix it, especially if it was working around a core limitation

**Comment by mawinter69** [comment]
in the config.jelly files of the verifiers in https://github.com/jenkinsci/ssh-slaves-plugin/tree/main/src/main/resources/hudson/plugins/sshslaves/verifiers

**Comment by timja** [comment]
> in the config.jelly files of the verifiers in [https://github.com/jenkinsci/ssh-slaves-plugin/tree/main/src/main/resources/hudson/plugins/sshslaves/verifiers](https://github.com/jenkinsci/ssh-slaves-plugin/tree/main/src/main/resources/hudson/plugins/sshslaves/verifiers?rgh-link-date=2022-02-28T15%3A05%3A30Z)

it documents it as a hack:
https://github.com/jenkinsci/ssh-slaves-plugin/blob/f3105af5e5d3adf9343c74ebae03c08b6c2fafe4/src/main/resources/hudson/plugins/sshslaves/verifiers/ManuallyProvidedKeyVerificationStrategy/config.jelly#L27

Could you send a PR to remove that?

**Comment by timja** [comment]
The help icon is orphaned on it's own here:

image

I think it needs the display name or some other label there otherwise it looks really strange

**Comment by mawinter69** [comment]
Indeed it doesn't look nice, but repeating what was just written in the dropdown is also not so nice. I would have said put the question mark for the help to the right of the drop down, but this would make the looks inconsistent with other drop downs probably.

**Comment by timja** [comment]
fyi https://github.com/jenkinsci/jenkins/pull/5417 is merged. could you merge in master?

**Review by timja — APPROVED** [review]
nice work!

File: core/src/main/resources/hudson/model/Slave/help-retentionStrategy.jelly
Comment (NotMyFault) [inline core/src/main/resources/hudson/model/Slave/help-retentionStrategy.jelly:26]:
```suggestion
<j:jelly xmlns:j="jelly:core" xmlns:l="/lib/layout">
```

File: core/src/main/resources/hudson/slaves/RetentionStrategy/Always/help.jelly
Comment (NotMyFault) [inline core/src/main/resources/hudson/slaves/RetentionStrategy/Always/help.jelly:2]:
```suggestion
<j:jelly xmlns:j="jelly:core">
```

File: core/src/main/resources/hudson/slaves/RetentionStrategy/Demand/help.jelly
Comment (NotMyFault) [inline core/src/main/resources/hudson/slaves/RetentionStrategy/Demand/help.jelly:2]:
```suggestion
<j:jelly xmlns:j="jelly:core">
```

File: core/src/main/resources/hudson/slaves/SimpleScheduledRetentionStrategy/help.jelly
Comment (NotMyFault) [inline core/src/main/resources/hudson/slaves/SimpleScheduledRetentionStrategy/help.jelly:2]:
```suggestion
<j:jelly xmlns:j="jelly:core">
```

**Review by NotMyFault — COMMENTED** [review]
Looks much better now

**Comment by timja** [comment]
Hmm it looks like this highlights an issue similar to https://github.com/jenkinsci/jenkins/pull/4562

Azure retention strategy items are showing up in the help whereas they didn't before this PR:

image


(on 2.342)
image

It's using a `DescriptorVisibilityFilter` to hide it:

https://github.com/jenkinsci/azure-vm-agents-plugin/blob/master/src/main/java/com/microsoft/azure/vmagent/AzureVMCloudPoolRetentionStrategy.java#L188-L194

I've checked both the Kubernetes plugin and EC2 plugin.

Kubernetes plugin seems to have created it's own retention strategy not based off the core one.
EC2 plugin doesn't seem to use it either, there's a class for it but seems like all the options are available as part of the template config, which kinda makes sense there was a PR here which did something similar except just duplicated a lot of options rather than flattening it: https://github.com/jenkinsci/azure-vm-agents-plugin/pull/183

----

Not sure if we should proceed right now, anyone got any thoughts?

**Comment by mawinter69** [comment]
I think I fixed this now 
Before it was looping over the filtered list, so I just made it the same.
EC2 plugin has no Extension on their descriptor so it is not shown.
Same for other retentionstrategies.

**Comment by timja** [comment]
List is empty for me on the configure agent screen of a permanent agent:

image

**Comment by timja** [comment]
Fixed for you in https://github.com/jenkinsci/jenkins/pull/6255/commits/eec7a460eeca4ed647cf0ff89297936f5f09fe0b

image

**Review by NotMyFault — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
